src/main/java/dev/openfeature/sdk/EventEmitter.java

```diff
@@ -0,0 +1,108 @@
+    private AutoCloseableReentrantReadWriteLock handlersLock = new AutoCloseableReentrantReadWriteLock();
+
+    /**
+     * Construct a new EventEmitter.
+     */
+    public EventEmitter() {
+        handlerMap = new ConcurrentHashMap<ProviderEvent, List<Consumer<EventDetails>>>();
+        handlerMap.put(ProviderEvent.PROVIDER_READY, new ArrayList<>());
+        handlerMap.put(ProviderEvent.PROVIDER_CONFIGURATION_CHANGED, new ArrayList<>());
+        handlerMap.put(ProviderEvent.PROVIDER_ERROR, new ArrayList<>());
+        handlerMap.put(ProviderEvent.PROVIDER_STALE, new ArrayList<>());
+    }
+
+    /**
+     * Emit an event.
+     *
+     * @param event   Event type to emit.
+     * @param details Event details.
+     */
+    public void emit(ProviderEvent event, ProviderEventDetails details) {
+        try (AutoCloseableLock __ = handlersLock.readLockAutoCloseable()) {
+            EventDetails eventDetails = EventDetails.builder()
+                    .flagMetadata(details.getFlagMetadata())
+                    .flagsChanged(details.getFlagsChanged())
+                    .message(details.getMessage())
+                    .build();
+
+            // we may be forwarding this event, preserve the name if so.
+            if (EventDetails.class.isInstance(details)) {
+                eventDetails.setClientName(((EventDetails) details).getClientName());
+            }
+
+            this.handlerMap.get(event).stream().forEach(handler -> {
+                runHander(handler, eventDetails);
+            });
+        }
+    }
+
+    void runHander(Consumer<EventDetails> handler, EventDetails eventDetails) {
+        taskExecutor.submit(() -> {
+            try {
+                handler.accept(eventDetails);
+            } catch (Exception e) {
+                log.error("Exception in event handler {}", handler, e);
+            }
+        });
+    }
+
+    void addHandler(ProviderEvent event, Consumer<EventDetails> handler) {
+        try (AutoCloseableLock __ = handlersLock.writeLockAutoCloseable()) {
+            this.handlerMap.get(event).add(handler);
+        }
+    }
+
+    void removeHandler(ProviderEvent event, Consumer<EventDetails> handler) {
+        try (AutoCloseableLock __ = handlersLock.writeLockAutoCloseable()) {
+            this.handlerMap.get(event).remove(handler);
+        }
+    }
+
+    void removeAllHandlers() {
+        try (AutoCloseableLock __ = handlersLock.writeLockAutoCloseable()) {
+            this.handlerMap.keySet().stream()
+                .forEach(type -> handlerMap.get(type).clear());
+        }
+    }
+
+    /**
+     * Propagates all events from the originatingEmitter to this one.
+     *
+     * @param originatingEmitter The emitter to forward events from.
+     * @param clientName         The client name that will be added to the events.
+     */
+    void forwardEvents(EventEmitter originatingEmitter, @Nullable String clientName) {
```

Comment:
This method allows us to "proxy" events from the provider in question, and easily move them when the provider is updated.